### PR TITLE
fix: make sure traffic source non-null

### PR DESCRIPTION
**Pull request by amliuyong**
----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

## Summary

(describe what this merge request does)

## Implementation highlights

(describe how the merge request does for feature changes, share the RFC link if it has)

## Test checklist

- [x] add new test cases
- [ ] all code changes are covered by unit tests
- [ ] end-to-end tests
  - [ ] deploy web console with CloudFront + S3 + API gateway
  - [ ] deploy web console within VPC
  - [ ] deploy ingestion server
    - [ ] with MSK sink
    - [ ] with KDS sink
    - [ ] with S3 sink
  - [ ] deploy data processing
  - [ ] deploy data modeling
    - [ ] new Redshift Serverless
    - [ ] provisioned Redshift
    - [ ] Athena
  - [ ] deploy with reporting
  - [ ] streaming ingestion
    - [ ] with Redshift Serverless
    - [ ] with provisioned Redshift

## Is it a breaking change

- [ ] add parameters without default value in stack
- [ ] introduce new service permission in stack
- [ ] introduce new top level stack module

## Miscellaneous

- [ ] introduce new symbol link source file(s) to be shared among infra code, web console frontend, and web console backend

### CODE REVIEW

**Comment by zxkane** [comment]
# SonarQube Code Analytics
## Quality Gate passed

passed

## Additional information
*The following metrics might not affect the Quality Gate status but improving them will improve your project code quality.*

## Issues
bug   [0 Bugs](http://sonarqube:9000/project/issues?id=pr-1397&resolved=false&sinceLeakPeriod=true&types=BUG)

vulnerability   [0 Vulnerabilities](http://sonarqube:9000/project/issues?id=pr-1397&resolved=false&sinceLeakPeriod=true&types=VULNERABILITY)

code_smell   [0 Code Smells](http://sonarqube:9000/project/issues?id=pr-1397&resolved=false&sinceLeakPeriod=true&types=CODE_SMELL)

## Coverage and Duplications
*No data* **Coverage**

*No data* **Duplication**

File: src/data-pipeline/spark-etl/src/main/java/software/aws/solution/clickstream/transformer/BaseTransformerV3.java [hunk 1]
```diff
@@ -84,6 +85,7 @@ public Dataset<Row> extractEvent(final Dataset<Row> convertedDataset) {
         List<String> allFields = ModelV2.getEventFields();
         Dataset<Row> eventDataset = convertedDataset.select(explode(expr("dataOut.events")).alias("event"))
                 .select("event.*")
+                .withColumn(Constant.TRAFFIC_SOURCE_SOURCE, coalesce(col(Constant.TRAFFIC_SOURCE_SOURCE), lit(DIRECT)))
```
Comment (zxkane) [inline src/data-pipeline/spark-etl/src/main/java/software/aws/solution/clickstream/transformer/BaseTransformerV3.java:88]: any requirement for assigning it to DIRECT?

we could treat null as unknown while disabling traffic source parsing, the value DIRECT has different meanings.

File: src/data-pipeline/etl-common/src/main/java/software/aws/solution/clickstream/common/sensors/SensorsEventParser.java [hunk 2]
```diff
@@ -59,25 +59,20 @@ public final class SensorsEventParser extends BaseEventParser {
...
+        return newInstance(null);
     }
 
-    public static SensorsEventParser getInstance(final TransformConfig transformConfig) {
-        if (instance == null) {
-            instance = new SensorsEventParser(transformConfig);
-        }
-        return instance;
+    public static SensorsEventParser newInstance(final TransformConfig transformConfig) {
+        return new SensorsEventParser(transformConfig);
```
Comment (zxkane) [inline src/data-pipeline/etl-common/src/main/java/software/aws/solution/clickstream/common/sensors/SensorsEventParser.java:73]: any benefits? for thread safe?

**Comment by amliuyong** [comment]
PR https://github.com/awslabs/clickstream-analytics-on-aws/pull/1399 will handle null value

**Pull request closed without merging**